phpseclib/File/ANSI.php

```diff
@@ -445,12 +395,32 @@ function appendString($source)
                 case "\n":
                     $this->_newLine();
                     break;
+                case "\x08": // backspace
+                    if ($this->x) {
+                        $this->x--;
+                        $this->attrs[$this->y][$this->x] = clone($this->base_attr_cell);
+                        $this->screen[$this->y] = substr_replace(
+                            $this->screen[$this->y],
+                            $source[$i],
+                            $this->x,
+                            1
+                        );
+                    }
+                    break;
                 case "\x0F": // shift
                     break;
                 case "\x1B": // start ANSI escape code
+                    $this->tokenization[count($this->tokenization) - 1] = substr($this->tokenization[count($this->tokenization) - 1], 0, -1);
+                    //if (!strlen($this->tokenization[count($this->tokenization) - 1])) {
```

Comment:
It's useful for diagnostic purposes. If I want to see what ANSI escape codes are being utilized I can either try to figure that out by hand or I can utilize the built in ANSI parsing routines in this class and have it spit them out to me, interleaved with the non-ANSI escape codes.

It's not documented anywhere and doesn't have a public API method for it. Pretty much it's not a feature that's of much benefit unless you're intending on developing this file. I suppose I could remove it along with the user_error's and put them in some sort of private notes of mine but that would make my life harder cause I'd have to consult those notes to copy / paste the code (it's easier to remember that I have `$this->tokenization` in the file than it is for me to remember that whole line) and it's not like removing that line would yield significant performance gains or anything.

So, basically, it's not hurting anyone and it's helping me. And the fact that it isn't documented means that it's not an officially supported feature and that semantic versioning can be maintained without doing a new "major" release if if this "feature" is removed or changed.